src/log/file_appender.cpp

```diff
@@ -82,13 +82,16 @@ namespace fc {
                    out.flush();
                    out.close();
                }
+               string rez;
+               read_file_contents(link_filename, rez);
                remove_all(link_filename);  // on windows, you can't delete the link while the underlying file is opened for writing
-               if( fc::exists( log_filename ) )
-                  out.open( log_filename, std::ios_base::out | std::ios_base::app );
-               else
-                  out.open( log_filename, std::ios_base::out | std::ios_base::app);
-
+               bool flag = false;
+               if (fc::exists( log_filename ))
+                    flag = true;
+               out.open( log_filename, std::ios_base::out | std::ios_base::app );
```

Comment:
I think the actual bug is inside `out.open(...)`, it is called here with `std::ios_base::app`, technically it should open the file in append mode, but it actually opened the file in overwrite mode.

Reply:
after making next changes in fstream.cpp file
void ofstream::open( const fc::path& file, int m ) {
     const boost::filesystem::path& bfp = file; 
     my->ofs.open( bfp, (std::ios_base::openmode)m | std::ios::binary );
   }

I see next compile errors, can anybody help me please ? @abitmore @oxarbitrage @jmjatlanta @ryanRfox @clockworkgr , I think there is also another issue, what do you think guys ?

[ 73%] Generating egenesis_brief.cpp, egenesis_full.cpp
embed_genesis:  Reading genesis from file /home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/genesis.json

embed_genesis:  parsing tmplsub parameter "/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_brief.cpp.tmpl---/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_brief.cpp"
embed_genesis:  parsing tmplsub parameter "/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp.tmpl---/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp"
Scanning dependencies of target graphene_egenesis_full
[ 74%] Building CXX object libraries/egenesis/CMakeFiles/graphene_egenesis_full.dir/egenesis_full.cpp.o
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:941870:50: error: redefinition of ‘const char graphene::egenesis::genesis_json_array [941747][41]’
 static const char genesis_json_array[941747][40+1] =
                                                  ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:49:19: note: ‘const char graphene::egenesis::genesis_json_array [941747][41]’ previously defined here
 static const char genesis_json_array[941747][40+1] =
                   ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp: In function ‘graphene::chain::chain_id_type graphene::egenesis::get_egenesis_chain_id()’:
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:1883621:15: error: redefinition of ‘graphene::chain::chain_id_type graphene::egenesis::get_egenesis_chain_id()’
 chain_id_type get_egenesis_chain_id()
               ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:941800:15: note: ‘graphene::chain::chain_id_type graphene::egenesis::get_egenesis_chain_id()’ previously defined here
 chain_id_type get_egenesis_chain_id()
               ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp: In function ‘void graphene::egenesis::compute_egenesis_json(std::__cxx11::string&)’:
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:1883626:6: error: redefinition of ‘void graphene::egenesis::compute_egenesis_json(std::__cxx11::string&)’
 void compute_egenesis_json( std::string& result )
      ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:941805:6: note: ‘void graphene::egenesis::compute_egenesis_json(std::__cxx11::string&)’ previously defined here
 void compute_egenesis_json( std::string& result )
      ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp: In function ‘fc::sha256 graphene::egenesis::get_egenesis_json_hash()’:
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:1883637:12: error: redefinition of ‘fc::sha256 graphene::egenesis::get_egenesis_json_hash()’
 fc::sha256 get_egenesis_json_hash()
            ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:941816:12: note: ‘fc::sha256 graphene::egenesis::get_egenesis_json_hash()’ previously defined here
 fc::sha256 get_egenesis_json_hash()
            ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp: At global scope:
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:1883691:50: error: redefinition of ‘const char graphene::egenesis::genesis_json_array [941747][41]’
 static const char genesis_json_array[941747][40+1] =
                                                  ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:49:19: note: ‘const char graphene::egenesis::genesis_json_array [941747][41]’ previously defined here
 static const char genesis_json_array[941747][40+1] =
                   ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp: In function ‘graphene::chain::chain_id_type graphene::egenesis::get_egenesis_chain_id()’:
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:2825442:15: error: redefinition of ‘graphene::chain::chain_id_type graphene::egenesis::get_egenesis_chain_id()’
 chain_id_type get_egenesis_chain_id()
               ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:941800:15: note: ‘graphene::chain::chain_id_type graphene::egenesis::get_egenesis_chain_id()’ previously defined here
 chain_id_type get_egenesis_chain_id()
               ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp: In function ‘void graphene::egenesis::compute_egenesis_json(std::__cxx11::string&)’:
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:2825447:6: error: redefinition of ‘void graphene::egenesis::compute_egenesis_json(std::__cxx11::string&)’
 void compute_egenesis_json( std::string& result )
      ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:941805:6: note: ‘void graphene::egenesis::compute_egenesis_json(std::__cxx11::string&)’ previously defined here
 void compute_egenesis_json( std::string& result )
      ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp: In function ‘fc::sha256 graphene::egenesis::get_egenesis_json_hash()’:
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:2825458:12: error: redefinition of ‘fc::sha256 graphene::egenesis::get_egenesis_json_hash()’
 fc::sha256 get_egenesis_json_hash()
            ^
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/libraries/egenesis/egenesis_full.cpp:941816:12: note: ‘fc::sha256 graphene::egenesis::get_egenesis_json_hash()’ previously defined here
 fc::sha256 get_egenesis_json_hash()
            ^
libraries/egenesis/CMakeFiles/graphene_egenesis_full.dir/build.make:72: recipe for target 'libraries/egenesis/CMakeFiles/graphene_egenesis_full.dir/egenesis_full.cpp.o' failed
make[2]: *** [libraries/egenesis/CMakeFiles/graphene_egenesis_full.dir/egenesis_full.cpp.o] Error 1
CMakeFiles/Makefile2:929: recipe for target 'libraries/egenesis/CMakeFiles/graphene_egenesis_full.dir/all' failed
make[1]: *** [libraries/egenesis/CMakeFiles/graphene_egenesis_full.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2

Reply:
I don't think the compiler error is related to the change.

Reply:
Oh I guess I was wrong. It can be related. The temporary/intermediate genesis file are opened with append mode. Need to fix it.

Reply:
The compiler error is due to [this code](https://github.com/bitshares/bitshares-core/blob/af118f697319d473b2547e2b110d0493980ac2b4/libraries/egenesis/embed_genesis.cpp#L283-L285):
```
      fc::ofstream outfile( dest_filename );
      outfile.write( out_str.c_str(), out_str.size() );
      outfile.close();
```

It turns out that the 2nd parameter `int m` of `ofsteam::open()` is incompatible with `std::ios_base::openmode`, see the definition here:
https://github.com/bitshares/bitshares-fc/blob/c16bb206a39b09b03423088c831bc193de89056c/include/fc/io/fstream.hpp#L10-L15

This is horrible coding practice, it may have been impacting a lot of code elsewhere, please be extremely careful. We may need to write new test cases for all of them.

@pmconrad @jmjatlanta please take a look if have time.

Reply:
should I fix  the temporary/intermediate genesis file to be opened with binary mode ? right ? or who should fix it ? can I do it ? if YES, can you help me please to find where are those code lines in which files ?

Reply:
void ofstream::open( const fc::path& file, int m ) {
     const boost::filesystem::path& bfp = file; 
     my->ofs.open( bfp, (std::ios_base::openmode)m );
   }

Now this code compiled fully and without any errors, strange :-) cannot understand the mystics, researching more and more deeply code ... I'm trying to fix this issue with the best quality and with only required minimal amount of changes !

Reply:
Commented here: https://github.com/bitshares/bitshares-core/issues/809#issuecomment-402428470